Fuzzing framework support

furszy opened the pull request:
As the title says, adding fuzzing framework support so we can start getting serious on this area as well.

Adapted the following PRs:

* #9172.
* #9354.
* #9691.
* #10415.
* #10440.
* #15043.
* #15047.
* #15295.
* #15399 (fabcfa5f0cfe2484fe5cf5ced7cbae34a35c47c9 only).
* #16338.
* #17051.
* #17076.
* #17225.
* #17942.
* #16236 (only fa35c4239f9b5a35ed46d0af8472c74610cba5c4).
* #18166 (only f2472f64604a0c583f950c56e8753d0bee246388).
* #18300.
* And.. probably will go further and continue adapting more PRs..

furszy commented:
Short update,
 Went further down upstream's path, improved the `fuzzing.md` document, disabled other unneeded targets when fuzz is enabled, added a python test_runner to run the fuzzer simpler `test/fuzz/test_runner.py` (testing needed..), added fuzzing harness for `CheckTransaction()`, `IsStandard()` and several `CTransaction` methods, plus added new cases to the deserialization targets.

As the changes are pretty isolated from everything else, only touching the new fuzzing framework, this PR shouldn't be too hard to review even when it has a good number of commits.

random-zebra commented on .travis/test_06_script_b.sh:
```diff
@@ -26,4 +26,10 @@ if [ "$RUN_FUNCTIONAL_TESTS" = "true" ]; then
   END_FOLD
 fi
 
+if [ "$RUN_FUZZ_TESTS" = "true" ]; then
+  BEGIN_FOLD fuzz-tests
+  DOCKER_EXEC test/fuzz/test_runner.py -l DEBUG ${DIR_FUZZ_IN}
+  END_FOLD
+fi
```
Since we are no longer using Travis, it would be awesome to have the same thing for GA in the future.

random-zebra reviewed (commented):
This is looking awesome. Been testing with AFL for a while and all is good.
Still need to check libfuzzer and the python script.

Some initial feedback while going through the code: needs rebase on master (and slight rework) for the changes to `CheckTransaction`, and the documentation could use a bit of "pivx-fication".

furszy commented on src/test/fuzz/transaction.cpp:
```diff
@@ -0,0 +1,141 @@
+bool CheckTransactionAllFlagsDisabled(const CTransaction& tx, CValidationState state);
+bool CheckTransactionAllFlagsEnabled(const CTransaction& tx, CValidationState state);
+bool CheckTransactionColdStakingOnlyEnabled(const CTransaction& tx, CValidationState state);
+bool CheckTransactionSaplingEnabled(const CTransaction& tx, CValidationState state);
```
Cool yeah, good idea.

furszy commented:
Rebased on master + updated per feedback.

1) Updated `CheckTransaction` function call to the new function signature.
2) Added `ContextualCheckTransaction` function call to the transaction fuzz target.
3) Fuzzing documentation "PIVXified".

furszy commented:
Rebased on master so this work does not get behind.

random-zebra reviewed (approved):
ACK 0d3af08cc24c9980862d95c9fd4ec4caeec01029 (didn't complete all the inputs... it's going too slow here).

Great job :tumbler_glass: 
I think this can be merged already, as it is a completely separated module, not touching the core code directly. Other upgrades can be done in successive pull requests.

Fuzzbawls commented:
I'm getting a linker error when building with `--enable-fuzz` using gcc, afl, and clang. Changing the link order as follows fixes it up for me on linux, but haven't tested macOS:
```diff
Index: src/Makefile.test.include
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/Makefile.test.include b/src/Makefile.test.include
--- a/src/Makefile.test.include	(revision 5f0ae0f7f4863ec6c75f8124b5331403003df9b3)
+++ b/src/Makefile.test.include	(date 1620471251595)
@@ -198,7 +198,7 @@
 if ENABLE_ZMQ
 test_test_pivx_fuzzy_LDADD += $(LIBBITCOIN_ZMQ) $(ZMQ_LIBS)
 endif
-test_test_pivx_fuzzy_LDADD += $(LIBBITCOIN_SERVER) $(LIBBITCOIN_CLI) $(LIBBITCOIN_COMMON) $(LIBBITCOIN_UTIL) $(LIBUNIVALUE) $(LIBBITCOIN_ZEROCOIN) \
+test_test_pivx_fuzzy_LDADD += $(LIBBITCOIN_SERVER) $(LIBBITCOIN_CLI) $(LIBBITCOIN_COMMON) $(LIBBITCOIN_ZEROCOIN) $(LIBBITCOIN_UTIL) $(LIBUNIVALUE) \
   $(LIBLEVELDB) $(LIBLEVELDB_SSE42) $(LIBMEMENV) $(BOOST_LIBS) $(LIBSECP256K1) $(EVENT_LIBS) $(EVENT_PTHREADS_LIBS)
 if ENABLE_WALLET
 test_test_pivx_fuzzy_LDADD += $(LIBBITCOIN_WALLET)

```

Other than this, I agree it could be merged sooner rather than later. Also, do we want to eventually maintain our own clone of the `qa-assets` repo? as surely our inputs will diverge from upstream.

furszy commented:
Hmm that is weird, i have been able to compile the fuzz target in macOS and linux (clang and gcc) fine all of this time and zebra too. But can add the link order change, compiled fine with it as well.
------------

About the `qa-assets` repo, yeah. We differ from upstream seed inputs already due the fact that we don't share the same data model. The serialization/deserialization and validation are different (transactions, block headers, block signature, inventory msgs include tier two types etc).

i'm planning to push the new repo when we finish 2267 work list, the modifications and news included there make some of the current inputs invalid which will force us to restart the entire fuzzing process to generate the correct inputs (which is so so slow and cpu consuming that is better to do it once the new changes are settled).

furszy commented:
Rebased on master + included the libzerocoin link order change. Ready to go.
